solidity/contracts/converter/LiquidTokenConverter.sol

```diff
@@ -0,0 +1,287 @@
+      * @dev creates a new converter with the given arguments and transfers
+      * the ownership to the caller
+      * 
+      * @param _token             smart token governed by the converter
+      * @param _registry          address of a contract registry contract
+      * @param _maxConversionFee  maximum conversion fee, represented in ppm
+      * 
+      * @return a new converter
+    */
+    function createConverter(ISmartToken _token, IContractRegistry _registry, uint32 _maxConversionFee) public returns(IBancorConverter) {
+        BancorConverter converter = new LiquidTokenConverter(_token, _registry, _maxConversionFee);
+        converter.transferOwnership(msg.sender);
+        return converter;
+    }
+}
+
+/**
+  * @dev Liquid Token Converter
+  * 
+  * The liquid token converter is a specialized version of a converter that manages a liquid token.
+  *
+  * The converters govern a token with a single reserve and allow converting between the two.
+  * Liquid tokens usually have fractional reserve (reserve ratio smaller than 100%).
+*/
+contract LiquidTokenConverter is BancorConverter {
+    /**
+      * @dev initializes a new LiquidTokenConverter instance
+      * 
+      * @param  _token              liquid token governed by the converter
+      * @param  _registry           address of a contract registry contract
+      * @param  _maxConversionFee   maximum conversion fee, represented in ppm
+    */
+    constructor(
+        ISmartToken _token,
+        IContractRegistry _registry,
+        uint32 _maxConversionFee
+    )
+        BancorConverter(_token, _registry, _maxConversionFee)
+        public
+    {
+    }
+
+    /**
+      * @dev returns the converter type
+      *
+      * @return see the converter types in the the main contract doc
+    */
+    function converterType() public pure returns (uint8) {
+        return 0;
+    }
+
+    /**
+      * @dev defines the reserve token for the converter
+      * can only be called by the owner while the converter is inactive and the
+      * reserve wasn't defined yet
+      * 
+      * @param _token   address of the reserve token
+      * @param _weight  reserve weight, represented in ppm, 1-1000000
+    */
+    function addReserve(IERC20Token _token, uint32 _weight) public {
+        // verify that the converter doesn't have a reserve yet
+        require(reserveTokenCount() == 0);
+        super.addReserve(_token, _weight);
+    }
+
+    /**
+      * @dev returns the expected rate of converting the source token to the
+      * target token along with the fee
+      * 
+      * @param _sourceToken contract address of the source token
+      * @param _targetToken contract address of the target token
+      * @param _amount      amount of tokens received from the user
+      * 
+      * @return expected rate
+      * @return expected fee
+    */
+    function rateAndFee(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount) public view returns (uint256, uint256) {

```

Comment:
@yudilevi:

There are 4 cases in this function:
1. source token == token && target token == token (invalid)
2. source token != token && target token == token (valid)
3. source token == token && target token != token (valid)
4. source token != token && target token != token (invalid)

If you look at it this way, then the `_sourceToken != _targetToken` check is a kind of a mixture of cases 1 and 4, which we can get rid of if we handle cases 2 and 3 explicitly.

I therefore recommend a different scheme, which:
1. Is more explicit about the exact type of each case of valid input
2. Handles all cases of valid input at the beginning, and then reverts for all the other cases

Here it is:
```
        if (_sourceToken != token && _targetToken == token)
            return purchaseRate(_amount);
        if (_sourceToken == token && _targetToken != token)
            return saleRate(_amount);

        // invalid input
        revert();
```

Reply:
updated - something like that?

Reply:
@yudilevi : Done?

Reply:
yeah I changed the code, take a look